core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +120,146 @@ where
             hasher,
             hashes,
             binary_tree_size,
+            start_index: 0,
+            cache: vec![None; depth],
         }
     }
 
+    /// Returns `true` if the tree is empty.
+    pub fn is_empty(&self) -> bool {
+        self.start_index == 0 && self.hashes.is_empty()
+    }
+
     /// Returns the root hash of this tree.
     /// # Panics
-    /// Will panic if the constant below is invalid.
-    pub fn merkle_root(self) -> H256 {
+    /// Should never panic, unless there is a bug.
+    pub fn merkle_root(&self) -> H256 {
         if self.hashes.is_empty() {
             let depth = tree_depth_by_size(self.binary_tree_size);
-            self.hasher.empty_subtree_hash(depth)
-        } else {
-            self.compute_merkle_root_and_path(0, None)
+            if self.start_index == 0 {
+                return self.hasher.empty_subtree_hash(depth);
+            } else if self.start_index == self.binary_tree_size {
+                return self.cache[depth].expect("cache is invalid");
+            }
         }
+        self.compute_merkle_root_and_path(0, None, None)
     }
 
     /// Returns the root hash and the Merkle proof for a leaf with the specified 0-based `index`.
-    pub fn merkle_root_and_path(self, index: usize) -> (H256, Vec<H256>) {
-        let mut merkle_path = vec![];
-        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut merkle_path));
-        (root_hash, merkle_path)
+    /// `index` is relative to the leftmost uncached leaf.
+    /// # Panics
+    /// Panics if `index` is >= than the number of leaves in the tree.
+    pub fn merkle_root_and_path(&self, index: usize) -> (H256, Vec<H256>) {
+        assert!(index < self.hashes.len(), "leaf index out of bounds");
+        let mut end_path = vec![];
+        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut end_path), None);
+        (
+            root_hash,
+            end_path.into_iter().map(Option::unwrap).collect(),
```

Comment:
This awkwardness with `unwrap()`s makes me think that an alternative with filtering a full path after getting it from `compute_merkle_root_and_path()` would be better. I.e., `compute_merkle_root_and_path()` would have two args and would always set the full path; and then it could be filtered based on the side and index of the element that the path corresponds to.

Reply:
It won't always be the full path anyway, since if we try to fully trim a tree, the path is bound to have some `None`s anyway. And we agreed that using garbage/empty values instead of `None` is worse than explicitly signaling it with types.

Also, I think having `end_path.into_iter().map(Option::unwrap).collect()`
is not that bad and much more readable than having filtering like

```rust
fn filter_merkle_path_by_side(
    path: Vec<Option<H256>>,
    index: usize,
    side: Side,
) -> Vec<Option<H256>> {
    path.into_iter()
        .enumerate()
        .map(|(i, hash)| {
            if (index & (1 << i)) >> i == side as usize {
                None
            } else {
                hash
            }
        })
        .collect()
}
```

Reply:
> if we try to fully trim a tree, the path is bound to have some Nones anyway

Do you mean that the path would have values that cannot be cached (since they would be invalidated when new leaves are inserted into the tree)? If so, this doesn't contradict my proposal; in it, `compute_merkle_root_and_path` returns a current full path to the specified leaf in the tree, and it is caller's responsibility to decide what to do with it. So e.g., if the path is intended for caching, its right side needs to be removed.

Of course, it's subjective, but I like `filter_merkle_path_by_side()` better (maybe together with a proposal to have a special type for it); the only thing in your snippet I don't think makes sense is `path` type; should be `Vec<H256>` (i.e., a *full* path). IIUC, `compute_merkle_root_and_path()` can create full paths just by always inserting `hashes[index ^ 1]` to the path on each level; or am I missing something?